From 4273eab9acd8dd05aaf10f2d12472837c42caff9 Mon Sep 17 00:00:00 2001
From: Ioana Radulescu <ruxandra.radulescu@nxp.com>
Date: Wed, 13 Mar 2019 19:24:51 +0200
Subject: [PATCH 476/741] dpaa2-eth: Update pause frames settings on link up

commit 81c18ada6652169486c52d29c2042fdc7b541499 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Adjust Rx taildrop configuration for each change in pause
frame support.

Signed-off-by: Ioana Ciornei <ioana.ciornei@nxp.com>
Signed-off-by: Ioana Radulescu <ruxandra.radulescu@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/net/ethernet/freescale/dpaa2/dpaa2-eth.c | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/drivers/net/ethernet/freescale/dpaa2/dpaa2-eth.c b/drivers/net/ethernet/freescale/dpaa2/dpaa2-eth.c
index c1be2489f3c8..6eb0277f4812 100644
--- a/drivers/net/ethernet/freescale/dpaa2/dpaa2-eth.c
+++ b/drivers/net/ethernet/freescale/dpaa2/dpaa2-eth.c
@@ -1271,6 +1271,18 @@ static void disable_ch_napi(struct dpaa2_eth_priv *priv)
 
 static void update_tx_fqids(struct dpaa2_eth_priv *priv);
 
+static void update_pf(struct dpaa2_eth_priv *priv,
+		      struct dpni_link_state *state)
+{
+	bool pause_frames;
+
+	pause_frames = !!(state->options & DPNI_LINK_OPT_PAUSE);
+	if (priv->tx_pause_frames != pause_frames) {
+		priv->tx_pause_frames = pause_frames;
+		set_rx_taildrop(priv);
+	}
+}
+
 static int link_state_update(struct dpaa2_eth_priv *priv)
 {
 	struct dpni_link_state state = {0};
@@ -1290,6 +1302,7 @@ static int link_state_update(struct dpaa2_eth_priv *priv)
 	priv->link_state = state;
 	if (state.up) {
 		update_tx_fqids(priv);
+		update_pf(priv, &state);
 		netif_carrier_on(priv->net_dev);
 		netif_tx_start_all_queues(priv->net_dev);
 	} else {
-- 
2.17.1

